dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/impl/AggregateContext.kt

```diff
@@ -0,0 +1,135 @@
+
+    override fun <X> exchange(initial: X, body: (Field<X>) -> Field<X>): Field<X> {
+        val messages = messagesAt<X>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        return body(subject).also { field ->
+            val message = SingleOutboundMessage(field.localValue, field.excludeSelf())
+            val path = stack.currentPath()
+            check(!toBeSent.messages.containsKey(path)) {
+                """
+                    Alignment was broken by multiple aligned calls with the same path: $path.
+                    The most likely cause is an aggregate function call within a loop
+                """.trimIndent()
+            }
+            toBeSent = toBeSent.copy(messages = toBeSent.messages + (stack.currentPath() to message))
+            state = state + (stack.currentPath() to field.localValue)
+        }
+    }
+
+    override fun <Init, Ret> exchanging(
+        initial: Init,
+        body: YieldingContext<Field<Init>, Field<Ret>>.(Field<Init>) -> YieldingResult<Field<Init>, Field<Ret>>,
+    ): Field<Ret> {
+        val messages = messagesAt<Init>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        val context = YieldingContext<Field<Init>, Field<Ret>>()
+        var res: Option<YieldingResult<Field<Init>, Field<Ret>>>
+        body(context, subject).also {
+            res = it.some()
+            state = state + (stack.currentPath() to it.toReturn.localValue)
+        }
+        return res.getOrElse { error("This error should never be thrown") }.toReturn
+    }
+
+    override fun <Initial, Return> repeating(
+        initial: Initial,
+        transform: YieldingContext<Initial, Return>.(Initial) -> YieldingResult<Initial, Return>,
+    ): Return {
+        val context = YieldingContext<Initial, Return>()
+        var res: Option<YieldingResult<Initial, Return>>
+        transform(context, stateAt(stack.currentPath(), initial)).also {
+            res = it.some()
+            state = state + (stack.currentPath() to it.toReturn)
+        }
+        return res.getOrElse { error("This error should never be thrown") }.toReturn
+    }
+
+    override fun <Initial> repeat(
+        initial: Initial,
+        transform: (Initial) -> Initial,
+    ): Initial =
+        repeating(initial) {
+            val res = transform(it)
+            YieldingResult(res, res)
+        }
+
+    override fun <R> alignedOn(pivot: Any?, body: () -> R): R {
+        stack.alignRaw(pivot)
+        return body().also { stack.dealign() }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun <T> messagesAt(path: Path): Map<ID, T> = messages
+        .filter { it.messages.containsKey(path) }
+        .associate { it.senderId to it.messages[path] as T }
```

Comment:
```suggestion
        .mapNotNull { received ->
            received.messages.getOrDefault(path, NoEntry)
                .takeIf { it != NoEntry }
                ?.let { received.senderId to it }
        }
        .associate()

    private object NoEntry
```

Reply:
Your suggestion does not compile.
A similar working version would be:

```kotlin
@Suppress("UNCHECKED_CAST")
private fun <T> messagesAt(path: Path): Map<ID, T> = messages
    .mapNotNull { received ->
        received.messages.getOrElse(path) { NoValue }
            .takeIf { it != NoValue }
            ?.let { received.senderId to it as T }
    }
    .associate { it }    
```
I think it is more boilerplate and not so convenient compared to the actual implementation